src/enemy/cuccoAggr.c

```diff
@@ -0,0 +1,383 @@
+}
+
+void sub_08038DD8(CuccoAggrEntity* this) {
+    super->subAction = 1;
+    super->actionDelay = 0x10;
+    super->field_0xf = gUnk_080CF7BC[super->type];
+    COLLISION_OFF(super);
+    super->spritePriority.b1 = 0;
+    InitializeAnimation(super, 4);
+    EnqueueSFX(SFX_VO_CUCCO4);
+}
+
+void sub_08038E18(CuccoAggrEntity* this) {
+    if (super->field_0xf) {
+        if (--super->field_0xf == 0) {
+            sub_08079184();
+            sub_08039218(this);
+            COLLISION_ON(super);
+            super->spritePriority.b1 = 1;
+            super->spritePriority.b0 = 4;
+            super->direction = Random() & 0x1f;
+            ProcessMovement(super);
+            UpdateSpriteForCollisionLayer(super);
+        }
+    }
+
+    if (--super->actionDelay == 0) {
+        super->actionDelay = 0x10;
+        super->spriteSettings.flipX ^= (Random() & 1);
+        CuccoAggr_CreateFx(this);
+    }
+
+    GetNextFrame(super);
+}
+
+void sub_08038EAC(CuccoAggrEntity* this) {
+    super->spritePriority.b1 = 1;
+    sub_080390F8(this);
+    GetNextFrame(super);
+}
+
+void sub_08038ED0(CuccoAggrEntity* this) {
+    COLLISION_ON(super);
+    sub_08039218(this);
+}
+
+void sub_08038EE0(CuccoAggrEntity* this) {
+    ProcessMovement(super);
+    GetNextFrame(super);
+    sub_08039298(this);
+    if (GravityUpdate(super, 0x1C00) == 0) {
+        if (--super->actionDelay == 0) {
+            sub_0803901C(this);
+        } else {
+            super->zVelocity = 0x8000;
+        }
+    }
+}
+
+void sub_08038F20(CuccoAggrEntity* this) {
+    sub_08039298(this);
+    GetNextFrame(super);
+
+    if (super->frame & 0x80) {
+        sub_0803901C(this);
+    }
+}
+
+void sub_08038F44(CuccoAggrEntity* this) {
+    super->direction = GetFacingDirection(&gPlayerEntity, super);
+    sub_080390F8(this);
+    ProcessMovement(super);
+    sub_080044EC(super, 0x1800);
+    GetNextFrame(super);
+    sub_08039298(this);
+    if (--this->unk_78 == 0) {
+        sub_08039120(this);
+    } else if ((this->unk_78 & 0x1f) == 0) {
+        CuccoAggr_CreateFx(this);
+    }
+}
+
+void sub_08038FA0(CuccoAggrEntity* this) {
+    if ((super->actionDelay++ & 0x1f) == 0) {
+        CuccoAggr_CreateFx(this);
+    }
+
+    if (this->unk_78) {
+        if ((this->unk_78-- & 0x7) == 0) {
+            sub_08004596(super, GetFacingDirection(super, &gPlayerEntity));
+        }
+
+        sub_080390F8(this);
+        LinearMoveUpdate(super);
+        GetNextFrame(super);
+    } else {
+        LinearMoveUpdate(super);
+        GetNextFrame(super);
+        if (CuccoAggr_IsOutsideScroll(this)) {
+            DeleteEntity(super);
+        }
+    }
+}
+
+void sub_0803901C(CuccoAggrEntity* this) {
+    u32 r3 = Random();
+    u32 r5 = r3 & 3;
+
+    //! @bug This if was meant to check r5 instead of r3.
+    //  The result is that this if practically never happens and
+    //  there's a higher chance of the default branch of the next switch.
+    if (r3 == 3) {
```

Comment:
Is this really a bug? What happens to the behavior when it is removed?

Reply:
It checks whether the result of Random() is 3. That is ridiculously rare as to not affect r5 at all. The rest of the function is basically 'do mod 4 on rand and pick one of three according to whether the result is 1, 2 or (0,3), meaning a chances of 1/4,/1/4,1/2.
If the condition would have been on r5, assuming action is also unified, the resulting probability would be 1/3,/1/3,1/3.